Add git blame error reporting with notification

mmtftr opened the pull request:
<img width="1035" alt="Screenshot 2024-04-11 at 13 13 44" src="https://github.com/zed-industries/zed/assets/13402668/cd0e96a0-41c6-4757-8840-97d15a75c511">

Release Notes:

- Added a notification to show git blame errors.

Caveats:
- ~git blame now executes in foreground
executor  (required since the Fut is !Send)~

TODOs:
- After a failed toggle, the app thinks the blame
is shown. This means toggling again will do nothing
instead of retrying. (Caused by editor.show_git_blame
being set to true before the git blame is generated)
- ~(Maybe) Trim error?~ Done

mmtftr committed:
Add git blame error reporting with notification

Fixed stderr not being captured.
Added notification with the git blame error.
Added logging for git blame error.
Caveat: git blame now executes in foreground
executor  (required since the Fut is !Sync)

TODO: After a failed toggle, the app thinks the blame
is shown. This means toggling again will do nothing
instead of retrying. (Caused by editor.show_git_blame
being set to true before the git blame is generated)

osiewicz commented on crates/project/src/project.rs:
```diff
@@ -7708,15 +7709,29 @@ impl Project {
+                        .blame(&relative_path, content)
+                        .with_context(|| format!("Failed to git blame {relative_path:?}"));
+
+                    if let Err(ref err) = blame_res {
```
I'm not @mrnugget , but I guess forcing this to be on a foreground thread is a pretty big deal; wouldn't it work if we emitted a notification on error in the caller of `blame_buffer`?

mmtftr replied:
yeah, good point. did exactly that.

mmtftr committed:
fix test error from update within update

mmtftr committed:
fix caveat by moving error reporting

mrnugget commented on crates/editor/src/git/blame.rs:
```diff
@@ -253,10 +253,10 @@ impl GitBlame {
         let buffer_edits = self.buffer.update(cx, |buffer, _| buffer.subscribe());
         let snapshot = self.buffer.read(cx).snapshot();
-        let blame = self.project.read(cx).blame_buffer(&self.buffer, None, cx);
+        let blame = Project::blame_buffer(&self.project, &self.buffer, None, cx);
```
I'm not sure I'm following — why do we need this change?

mmtftr replied:
I initially needed it to get access to the Model and ModelContext objects but now I don't. I'll revert that part and push another commit

mrnugget commented:
More a higher-level question: will we display errors to the user that are actionable? In your screenshot, for example, is that something that the user can fix, or is that a bug on our side?

If it's not actionable for the user, should we instead just `.log_err()`?

mmtftr commented:
In my example, it's actionable. My git-blame-ignore-revs file was missing which caused an error with git blame. I could not figure this out until I manually ran zed from the console to see the error.

In the case that git blame fails for whatever reason, the user should see that it failed in my opinion. Without the notification, the experience is that the user toggles the git blame but nothing happens.

mrnugget commented:
> In my example, it's actionable. My git-blame-ignore-revs file was missing which caused an error with git blame. I could not figure this out until I manually ran zed from the console to see the error.

Okay, that makes sense. Yeah, I think if we can simplify the code more so that it only shows the error, that makes sense.

mmtftr commented:
> If it's not actionable for the user, should we instead just `.log_err()`?

I wanted to `log_err()` but for some reason that returns an Option, (I think it should be like inspect_err and not change the type).

mmtftr committed:
revert project blame_buffer api change

mmtftr committed:
add test coverage for git blame errors

mmtftr committed:
add context to the blame error again

mmtftr commented:
Added test coverage. Simplified the error context. The final error looks like this when the process fails: 
<img width="502" alt="Screenshot 2024-04-11 at 20 30 23" src="https://github.com/zed-industries/zed/assets/13402668/1c64c5a9-3c84-49b0-92a9-1041410e6e39">

mmtftr committed:
fix tests message

mrnugget committed:
Simplify code a little bit and make test more explicit

mrnugget commented:
Thanks. I hopped on the branch and simplified the code a little bit. Will merge once CI is through.
